File: app/src/main/java/hu/rayworks/vizit/ui/design/components/VizitBrandHeader.kt
package hu.rayworks.vizit.ui.design.components

import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.heading
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.outlined.ArrowBack
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.outlined.ChevronRight
import androidx.compose.ui.draw.clip
import androidx.compose.ui.semantics.Role
import hu.rayworks.vizit.ui.util.rememberProfilePhoto
import hu.rayworks.vizit.R
import hu.rayworks.vizit.ui.design.Vizit

private const val SLOGAN = "Egy érintés. Egy kapcsolat."

/**
 * How much brand presence a screen gets.
 *
 * [Full] is the brand moment on primary destinations: the mark, the VIZIT name
 * and the slogan, each at a size that is actually readable. [Compact] keeps all
 * three on secondary and inner screens — smaller, but never reduced to a
 * generic app bar where the brand disappears.
 */
enum class VizitBrandHeaderStyle(
    val markHeight: Dp,
    val nameSize: Int,
    val nameTracking: Double,
) {
    Full(markHeight = 34.dp, nameSize = 26, nameTracking = 5.0),
    Compact(markHeight = 22.dp, nameSize = 16, nameTracking = 3.0),
}

/**
 * The VIZIT mark on its mandated white plate.
 *
 * The artwork is a fixed brand asset: white background, blue "V". It is never
 * tinted or inverted, so the plate is part of the mark rather than decoration
 * around it. In dark mode a hairline keeps it from floating on the canvas.
 */
@Composable
fun VizitBrandMarkPlate(
    height: Dp,
    modifier: Modifier = Modifier,
) {
    val shape = RoundedCornerShape(Vizit.radius.md)
    Box(
        modifier = modifier
            .background(Color.White, shape)
            .border(
                1.dp,
                if (Vizit.colors.isDark) Color.White.copy(alpha = 0.16f) else Vizit.colors.border,
                shape,
            )
            .padding(horizontal = height * 0.34f, vertical = height * 0.26f),
        contentAlignment = Alignment.Center,
    ) {
        Image(
            painter = painterResource(R.drawable.vizit_logo_mark),
            contentDescription = null,
            contentScale = ContentScale.Fit,
            modifier = Modifier.height(height),
        )
    }
}

/**
 * The one brand header every screen uses. Screens never hand-roll their own.
 *
 * The caller supplies the window-inset padding, so this composable stays usable
 * inside scrolling content as well as directly under the status bar.
 */
@Composable
fun VizitBrandHeader(
    modifier: Modifier = Modifier,
    style: VizitBrandHeaderStyle = VizitBrandHeaderStyle.Full,
    onBack: (() -> Unit)? = null,
    trailing: @Composable (() -> Unit)? = null,
) {
    when (style) {
        VizitBrandHeaderStyle.Full -> Row(
            modifier = modifier.fillMaxWidth(),
            verticalAlignment = Alignment.Top,
        ) {
            Column(
                modifier = Modifier.weight(1f),
                verticalArrangement = Arrangement.spacedBy(Vizit.space.sm),
            ) {
                VizitBrandMarkPlate(height = style.markHeight)
                BrandWordmark(style)
            }
            trailing?.invoke()
        }

        VizitBrandHeaderStyle.Compact -> Row(
            modifier = modifier
                .fillMaxWidth()
                .defaultMinSize(minHeight = 48.dp),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(Vizit.space.sm),
        ) {
            if (onBack != null) {
                Box(
                    modifier = Modifier
                        .defaultMinSize(minWidth = 44.dp, minHeight = 44.dp)
                        .clickable(onClick = onBack)
                        .semantics { contentDescription = "Vissza" },
                    contentAlignment = Alignment.Center,
                ) {
                    Icon(
                        imageVector = Icons.AutoMirrored.Outlined.ArrowBack,
                        contentDescription = null,
                        tint = Vizit.colors.textSecondary,
                    )
                }
            }
            VizitBrandMarkPlate(height = style.markHeight)
            Box(modifier = Modifier.weight(1f)) { BrandWordmark(style) }
            trailing?.invoke()
        }
    }
}

@Composable
private fun BrandWordmark(style: VizitBrandHeaderStyle) {
    Column(
        modifier = Modifier.clearAndSetSemantics {
            contentDescription = "VIZIT – $SLOGAN"
            heading()
        },
    ) {
        Text(
            text = "VIZIT",
            style = Vizit.type.h1.copy(
                fontSize = style.nameSize.sp,
                letterSpacing = style.nameTracking.sp,
            ),
            color = Vizit.colors.textPrimary,
        )
        Text(
            text = SLOGAN,
            style = if (style == VizitBrandHeaderStyle.Full) Vizit.type.body else Vizit.type.caption,
            color = Vizit.colors.textSecondary,
            maxLines = 2,
            overflow = TextOverflow.Ellipsis,
        )
    }
}

/**
 * The signed-in person, shown as its own row under the brand rather than
 * competing with it. The real photo is used whenever there is one.
 */
@Composable
fun VizitUserBadge(
    displayName: String,
    initials: String,
    photoBase64: String,
    modifier: Modifier = Modifier,
    greeting: String = "Üdv újra,",
    onClick: (() -> Unit)? = null,
) {
    val colors = Vizit.colors
    val name = displayName.ifBlank { "Állítsd be a névjegyed" }
    val photo = rememberProfilePhoto(photoBase64)

    Row(
        modifier = modifier
            .fillMaxWidth()
            .then(if (onClick != null) Modifier.clickable(role = Role.Button, onClick = onClick) else Modifier)
            .defaultMinSize(minHeight = 48.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(Vizit.space.sm),
    ) {
        Box(
            modifier = Modifier
                .size(48.dp)
                .background(colors.primarySubtle, CircleShape)
                .border(1.dp, colors.border, CircleShape),
            contentAlignment = Alignment.Center,
        ) {
            if (photo != null) {
                Image(
                    bitmap = photo,
                    contentDescription = null,
                    contentScale = ContentScale.Crop,
                    modifier = Modifier.size(48.dp).clip(CircleShape),
                )
            } else {
                Text(
                    text = initials.ifBlank { "V" },
                    style = Vizit.type.label,
                    color = colors.primary,
                )
            }
        }

        Column(modifier = Modifier.weight(1f)) {
            Text(greeting, style = Vizit.type.bodySmall, color = colors.textMuted)
            Text(
                text = name,
                style = Vizit.type.h3,
                color = colors.textPrimary,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
            )
        }

        if (onClick != null) {
            Icon(
                imageVector = Icons.Outlined.ChevronRight,
                contentDescription = null,
                tint = colors.textMuted,
            )
        }
    }
}

/**
 * The screen's own title, rendered the way the platform renders one: large,
 * left-aligned, with an optional action or identity chip on the trailing edge.
 *
 * Primary screens carry the brand here instead of a separate header block, so
 * the brand costs nothing but the title line the screen already needed.
 */
@Composable
fun VizitLargeTitle(
    title: String,
    modifier: Modifier = Modifier,
    letterSpacing: Double = 0.0,
    trailing: @Composable (() -> Unit)? = null,
) {
    Row(
        modifier = modifier.fillMaxWidth().defaultMinSize(minHeight = 48.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(Vizit.space.sm),
    ) {
        Text(
            text = title,
            style = Vizit.type.display.copy(letterSpacing = letterSpacing.sp),
            color = Vizit.colors.textPrimary,
            maxLines = 1,
            overflow = TextOverflow.Ellipsis,
            modifier = Modifier.weight(1f).semantics { heading() },
        )
        trailing?.invoke()
    }
}

/**
 * The signed-in person as a tappable chip beside the title. It replaces the
 * full greeting row on the home screen: same destination, a fraction of the
 * height.
 */
@Composable
fun VizitIdentityChip(
    initials: String,
    photoBase64: String,
    displayName: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val colors = Vizit.colors
    val photo = rememberProfilePhoto(photoBase64)
    val label = if (displayName.isBlank()) {
        "Névjegy beállítása"
    } else {
        "Megnyitás: $displayName névjegye"
    }
    Box(
        modifier = modifier
            .size(44.dp)
            .clickable(role = Role.Button, onClick = onClick)
            .semantics { contentDescription = label },
        contentAlignment = Alignment.Center,
    ) {
        Box(
            modifier = Modifier
                .size(34.dp)
                .background(colors.primarySubtle, CircleShape)
                .border(1.dp, colors.border, CircleShape),
            contentAlignment = Alignment.Center,
        ) {
            if (photo != null) {
                Image(
                    bitmap = photo,
                    contentDescription = null,
                    contentScale = ContentScale.Crop,
                    modifier = Modifier.size(34.dp).clip(CircleShape),
                )
            } else {
                Text(
                    text = initials.ifBlank { "V" },
                    style = Vizit.type.caption,
                    color = colors.primary,
                )
            }
        }
    }
}

File: ios/Vizit/Design/VizitBrandHeader.swift
import SwiftUI

/// How much brand presence a screen gets.
///
/// `full` is the brand moment on primary destinations: the mark, the VIZIT
/// name and the slogan, each at a size that is actually readable. `compact`
/// keeps all three on secondary and inner screens — smaller, but never reduced
/// to a generic app bar where the brand disappears.
enum VizitBrandHeaderStyle {
    case full
    case compact

    var markHeight: CGFloat { self == .full ? 34 : 22 }
    var nameSize: CGFloat { self == .full ? 26 : 16 }
    var nameTracking: CGFloat { self == .full ? 5 : 3 }
    var sloganFont: Font { self == .full ? VizitFont.body : VizitFont.caption }
}

/// The VIZIT mark on its mandated white plate.
///
/// The artwork is a fixed brand asset: white background, blue "V". It is never
/// tinted, inverted or placed straight onto a themed surface, so the plate is
/// part of the mark rather than a decoration around it. In dark mode a hairline
/// keeps the plate from floating on the canvas.
struct VizitBrandMark: View {
    var height: CGFloat = 34

    private static let image: UIImage? = {
        guard let url = Bundle.main.url(forResource: "VizitLogoMark", withExtension: "png") else { return nil }
        return UIImage(contentsOfFile: url.path)
    }()

    var body: some View {
        Group {
            if let image = Self.image {
                Image(uiImage: image)
                    .resizable()
                    .scaledToFit()
                    .frame(height: height)
            } else {
                // The mark is the brand; if the asset is missing, the wordmark
                // still has to read as VIZIT rather than as an empty box.
                Text("V")
                    .font(.system(size: height * 0.9, weight: .black))
                    .foregroundStyle(VizitColor.primary)
                    .frame(height: height)
            }
        }
        .padding(.horizontal, height * 0.34)
        .padding(.vertical, height * 0.26)
        .background(Color.white)
        .clipShape(RoundedRectangle(cornerRadius: VizitRadius.md, style: .continuous))
        .overlay {
            RoundedRectangle(cornerRadius: VizitRadius.md, style: .continuous)
                .stroke(VizitColor.border, lineWidth: 1)
        }
        .accessibilityHidden(true)
    }
}

/// The one brand header every screen uses. Screens never hand-roll their own.
struct VizitBrandHeader<Trailing: View>: View {
    var style: VizitBrandHeaderStyle = .full
    var onBack: (() -> Void)?
    @ViewBuilder var trailing: () -> Trailing

    private let slogan = "Egy érintés. Egy kapcsolat."

    var body: some View {
        switch style {
        case .full: fullHeader
        case .compact: compactHeader
        }
    }

    private var wordmark: some View {
        Text("VIZIT")
            .font(.system(size: style.nameSize, weight: .bold))
            .tracking(style.nameTracking)
            .foregroundStyle(VizitColor.textPrimary)
    }

    private var sloganText: some View {
        Text(slogan)
            .font(style.sloganFont)
            .foregroundStyle(VizitColor.textSecondary)
            .lineLimit(2)
            .minimumScaleFactor(0.85)
            .fixedSize(horizontal: false, vertical: true)
    }

    private var fullHeader: some View {
        HStack(alignment: .top, spacing: VizitSpace.md) {
            VStack(alignment: .leading, spacing: VizitSpace.sm) {
                VizitBrandMark(height: style.markHeight)
                VStack(alignment: .leading, spacing: 2) {
                    wordmark
                    sloganText
                }
            }
            .accessibilityElement(children: .ignore)
            .accessibilityLabel("VIZIT – \(slogan)")
            .accessibilityAddTraits(.isHeader)

            Spacer(minLength: VizitSpace.sm)

            trailing()
        }
    }

    private var compactHeader: some View {
        HStack(spacing: VizitSpace.sm) {
            if let onBack {
                VizitIconButton(systemImage: "chevron.left", accessibilityTitle: "Vissza", action: onBack)
                    .padding(.leading, -VizitSpace.sm)
            }

            VizitBrandMark(height: style.markHeight)

            VStack(alignment: .leading, spacing: 0) {
                wordmark
                sloganText
            }
            .accessibilityElement(children: .ignore)
            .accessibilityLabel("VIZIT – \(slogan)")
            .accessibilityAddTraits(.isHeader)

            Spacer(minLength: VizitSpace.xs)

            trailing()
        }
        .frame(minHeight: VizitMetrics.minTouchTarget)
    }
}

extension VizitBrandHeader where Trailing == EmptyView {
    init(style: VizitBrandHeaderStyle = .full, onBack: (() -> Void)? = nil) {
        self.init(style: style, onBack: onBack) { EmptyView() }
    }
}

/// The signed-in person, shown as its own row under the brand rather than
/// competing with it. The real photo is used whenever there is one.
struct VizitUserBadge: View {
    let profile: ContactProfile
    var greeting = "Üdv újra,"
    /// Applied to the name itself, so UI tests can read it as a static text.
    /// Children are deliberately left un-combined for that reason.
    var nameIdentifier: String?
    var action: (() -> Void)?

    private var name: String {
        profile.displayName.isEmpty ? "Állítsd be a névjegyed" : profile.displayName
    }

    var body: some View {
        Button {
            action?()
        } label: {
            HStack(spacing: VizitSpace.sm) {
                VizitAvatar(profile: profile, size: 48)

                VStack(alignment: .leading, spacing: 1) {
                    Text(greeting)
                        .font(VizitFont.bodySmall)
                        .foregroundStyle(VizitColor.textMuted)
                    Text(name)
                        .font(VizitFont.h3)
                        .foregroundStyle(VizitColor.textPrimary)
                        .lineLimit(1)
                        .minimumScaleFactor(0.8)
                        .vizitIdentifier(nameIdentifier)
                }

                Spacer(minLength: 0)

                if action != nil {
                    Image(systemName: "chevron.right")
                        .font(.system(size: 14, weight: .semibold))
                        .foregroundStyle(VizitColor.textMuted)
                }
            }
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .disabled(action == nil)
        .accessibilityLabel("\(greeting) \(name)")
    }
}
